Add PVF execution priority

AndreiEres opened the pull request:
Resolves https://github.com/paritytech/polkadot-sdk/issues/4632

The new logic optimizes the distribution of execution jobs for disputes, approvals, and backings. Testing shows improved finality lag and candidate checking times, especially under heavy network load.

### Approach

This update adds prioritization to the PVF execution queue. The logic partially implements the suggestions from https://github.com/paritytech/polkadot-sdk/issues/4632#issuecomment-2209188695.

We use thresholds to determine how much a current priority can "steal" from lower ones:
-  Disputes: 70%
-  Approvals: 80%
-  Backing System Parachains: 100%
-  Backing: 100%

A threshold indicates the portion of the current priority that can be allocated from lower priorities.

For example:
-  Disputes take 70%, leaving 30% for approvals and all backings.
-  80% of the remaining goes to approvals, which is 30% * 80% = 24% of the original 100%.
-  If we used parts of the original 100%, approvals couldn't take more than 24%, even if there are no disputes.

Assuming a maximum of 12 executions per block, with a 6-second window, 2 CPU cores, and a 2-second run time, we get these distributions:

-  With disputes: 8 disputes, 3 approvals, 1 backing
-  Without disputes: 9 approvals, 3 backings

It's worth noting that when there are no disputes, if there's only one backing job, we continue processing approvals regardless of their fulfillment status.

### Versi Testing 40/20

Testing showed a slight difference in finality lag and candidate checking time between this pull request and its base on the master branch. The more loaded the network, the greater the observed difference.

Testing Parameters:
-  40 validators (4 malicious)
-  20 gluttons with 2 seconds of PVF execution time
-  6 VRF modulo samples
-  12 required approvals

![Pasted Graphic 3](https://github.com/user-attachments/assets/8b6163a4-a1c9-44c2-bdba-ce1ef4b1eba7)  
![Pasted Graphic 4](https://github.com/user-attachments/assets/9f016647-7727-42e8-afe9-04f303e6c862)

### Versi Testing 80/40

For this test, we compared the master branch with the branch from https://github.com/paritytech/polkadot-sdk/pull/5616. The second branch is based on the current one but removes backing jobs that have exceeded their time limits. We excluded malicious nodes to reduce noise from disputing and banning validators. The results show that, under the same load, nodes experience less finality lag and reduced recovery and check time. Even parachains are functioning with a shorter block time, although it remains over 6 seconds.

Testing Parameters:
-  80 validators (0 malicious)
-  40 gluttons with 2 seconds of PVF execution time
-  6 VRF modulo samples
-  30 required approvals

![image](https://github.com/user-attachments/assets/42bcc845-9115-4ae3-9910-286b77a60bbf)

sandreim commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +618,151 @@ pub fn start(
+
+struct Unscheduled {
+	unscheduled: HashMap<ExecutePriority, VecDeque<ExecuteJob>>,
+	priority_log: VecDeque<ExecutePriority>,
```
I think this can be implemented without the queue. We will count number of jobs dispatched per priority ( **A** - low, **B**- normal, **C** critical). Then, for every **N** jobs (in our case N = 10) received we reset the counters for each priority to 0. 

In your code we want to allow at least 10% of low priority jobs to be executed so `P = 10% * N, P > 0`. Meaning dispatch 1 low priority job out of every 10.

Then  `select_next_priority()` should return low priority only if:
- the normal/critical queues are empty
- `B + C == N - P`, meaning the normal/critical priority messages have used up all the allocated space.

AndreiEres commented:
Updated the logic according to https://github.com/paritytech/polkadot-sdk/issues/4632#issuecomment-2209188695 which looks reasonable

alexggh commented on polkadot/node/subsystem-types/src/messages.rs:
```diff
@@ -204,6 +204,50 @@ pub enum CandidateValidationMessage {
+/// Extends primitives::PvfExecKind to have a separate value for duspute requests
+/// which is important for prioritization.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum PvfExecPriority {
```
It is not clear to me what is the difference between this enum and PvfExecKind and why we can't extend PvfExecKind with a DisputeVariant, I assume, because ExecKind is used in the runtime as well ?

AndreiEres replied:
Yes, ExecKind is a runtime param and there is no need to change it. I would say using it further in all that code was a confusion because we needed only 2 options.

AndreiEres replied:
But I'll add a note here why we need it

alexggh commented on polkadot/node/subsystem-types/src/messages.rs:
```diff
@@ -204,6 +204,50 @@ pub enum CandidateValidationMessage {
+impl PvfExecPriority {
+	/// Returns an iterator over the variants of `PvfExecPriorityKind` in order
+	pub fn iter() -> impl Iterator<Item = PvfExecPriority> {
+		[Self::Dispute, Self::Approval, Self::BackingSystem, Self::Backing]
```
This won't age when you add new variants people will forget to add them here, you can use crates for this like: https://docs.rs/strum/latest/strum/derive.EnumIter.html.

alexggh commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +620,185 @@ pub fn start(
+
+impl Unscheduled {
+	// A threshold reaching which we prioritize approval jobs
+	const REGULAR_JOBS_THRESHOLD: usize = 12;
```
Nit: This is good enough, but if it doesn't require too much effort  I would derive from modulo_samples.

Also, please add a comment how we came up with this number.

alexggh commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +620,185 @@ pub fn start(
+			return Approval
+		}
+
+		if self.regular_jobs_count() > Self::REGULAR_JOBS_THRESHOLD && is_probable(87) {
```
I would add some metrics when this and the dispute path are hit.

alexggh reviewed (commented):
First pass, good job, I like the simplicity, I left you some comments, nothing major.

As we discussed online, it would be nice if we can run some integrations tests to understand how the system behaves on the limit conditions, I think we can simulate that by making the pvf execution longer, so that we fall behind on work.

sandreim commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +620,185 @@ pub fn start(
+// Return a bool with a probability of numerator/100 of being true,
+// which means the numerator can't be more than 100
+fn is_probable(numerator: u32) -> bool {
+	let mut rng = rand::thread_rng();
```
Yes, very good point. My suggestion to solve this from https://github.com/paritytech/polkadot-sdk/pull/4837#discussion_r1653157656 still stands. It can be tweaked to work fine with what is needed here.

alexggh commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -236,7 +243,7 @@ impl Queue {
 		if eldest.waiting_since.elapsed() < MAX_KEEP_WAITING {
```
We need to make it clear that this logic happens inside the same priority class wouldn't it ? 
It is not global anymore.

alexggh commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +622,181 @@ pub fn start(
+
+impl Unscheduled {
+	// A threshold reaching which we reset counted jobs.
+	// Max number of jobs per block assuming 6s window, 2 CPU cores, and 2s for a run.
```
```suggestion
	// The max expected queue_size in normal conditions, at the beginning of each block, 
	// a validator will submit jobs for at least vrf_module_samples(6) + 1 for backing
	// the parachain candidate they are assigned to, there is some buffer added to cover 
	// for situations, where more work arrives in the queue.
```

alexggh commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +622,181 @@ pub fn start(
+		(count + 1) * 100 / total_count >= threshold
+	}
+
+	fn log(&mut self, priority: PvfExecPriority) {
```
```suggestion
	fn mark_scheduled(&mut self, priority: PvfExecPriority) {
```

alexggh commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +622,181 @@ pub fn start(
+		*current_count += 1;
+
+		if self.counter.values().sum::<usize>() >= Self::MAX_COUNT {
+			self.reset_counter();
```
I think we discussed that if the queue is not above MAX_COUNT we would want to prioritize backing or at least backing for system chains, because backing is a bit more time sensitive than approval, the new refactor doesn't seem to cover that.

Are we ok with this trade-off ? If yes, can we find a way to properly log when this condition is met, so that we can at least detect it if it happens in production.

AndreiEres replied:
With the current logic with max loaded queues we should run a backing job once per 12 execution. Backing system parachains has more priority than just backing.

AndreiEres commented:
@alexggh Thank you for the good questions. I added a few test cases to show how the selection should work with different loads. Also added few debug lines.

alexggh commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -608,3 +622,271 @@ pub fn start(
+		let Some(threshold) = Self::fulfilled_threshold(priority) else { return false };
+		let Some(count) = self.counter.get(&priority) else { return false };
+		// Every time we iterate by lower level priorities
+		let total_count: usize = self
```
```suggestion
		let total_scheduled_at_prio_or_lower: usize = self
```

alexggh reviewed (approved):
Overall looks good, I would also run some long running networks(versi or zombienet) and check parachains produce blocks at 6s rate clip.

AndreiEres commented:
Testing results added to the description

sandreim commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -638,3 +653,289 @@ pub fn start(
+	// A threshold in percentages, the portion a current priority can "steal" from lower ones.
+	// For example:
+	// Disputes take 70%, leaving 30% for approvals and all backings.
+	// 80% of the remaining goes to approvals, which is 30% * 80% = 24% of the original 100%.
```
This way of expressing the percentages is not easy. You have to do this math everytime. Can it be made simpler to not require the reader to do much math ?

AndreiEres replied:
I tried to update the docs to make it clearer. Unfortunately I didn't come up with a better solution.

sandreim commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -638,3 +653,289 @@ pub fn start(
+}
+
+impl Unscheduled {
+	// A threshold reaching which we reset counted jobs.
+	// The max expected queue_size in normal conditions, at the beginning of each block,
+	// a validator will submit jobs for at least vrf_module_samples(6) + 1 for backing
+	// the parachain candidate they are assigned to, there is some buffer added to cover
+	// for situations, where more work arrives in the queue.
+	const MAX_COUNT: usize = 12;
```
This section is a bit confusing. Maybe `SCHEDULING_WINDOW_SIZE` is a better name.

sandreim commented on polkadot/node/core/candidate-validation/src/lib.rs:
```diff
@@ -993,6 +997,8 @@ trait ValidationBackend {
 		retry_delay: Duration,
 		// The priority for the preparation job.
 		prepare_priority: polkadot_node_core_pvf::Priority,
+		// The priority for the preparation job.
```
```suggestion
		// The priority for the execution job.
```

sandreim reviewed (approved):
Thank you @AndreiEres . This is very good work and the tests results show it. Just have a few more nits here and there, but otherwise good to go from my side.

@burdges any concerns with the percentages here: https://github.com/paritytech/polkadot-sdk/blob/e456a5a8a6afed1c5ad4ca2d4f524db5137ca85f/polkadot/node/core/pvf/src/execute/queue.rs#L673 ?

AndreiEres commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -638,3 +653,289 @@ pub fn start(
+		)
+	}
+
+	fn is_fulfilled(&self, priority: &PvfExecPriority) -> bool {
```
Changed the name, we have to check if we reached a threshold for the specified priority

burdges commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -638,3 +653,299 @@ pub fn start(
+	///
+	/// This system might seem complex, but we operate with the remaining percentages because:
+	/// - Not all job types are present in each block. If we used parts of the original 100%,
+	///   approvals could not exceed 24%, even if there are no disputes.
```
I've trouble parsing "even if there are no disputes" here.  If there are no backing jobs then approvals takes like 80% of the total, right?

AndreiEres replied:
Do you mean that we will skip jobs if we only have approvals in the queue and they have reached 80%? If every current priority has reached its limit, we will select the highest one to execute. In that case, we will continue picking approvals in the absence of other jobs.

burdges commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -638,3 +653,299 @@ pub fn start(
+	/// - Not all job types are present in each block. If we used parts of the original 100%,
+	///   approvals could not exceed 24%, even if there are no disputes.
+	/// - We cannot fully prioritize backing system parachains over backing other parachains based
+	///   on the distribution of the original 100%.
```
Yeah, we might've some gradiations among the different system parachains, so do whatever you think is best for now.  AssetHub could maybe halt without killing anything, although this depends upon how its used.  Among the chains that must run "enough" during an epoch:
- Sassafras tickets collection, if we ever moved it onto a parachain (maybe never).
- Validator elections should maintain a fresh validator set, even if we've lost many for some reason.
- If validator elections runs then we must run bridgehub enough, and any future DKG chains.
- Some governance chain without smart contracts (collectives?).

sandreim replied:
Yeah this is just a first step. I agree it needs to be e reconsidered once we start out to move governance and staking out of the relay chain.

However, I think that a better algorithm would actually consider to target specific block times for the system parachains in times of overload.

burdges replied:
Afaik, critical system paracahins should typically not require numerous blocks, specific block times, etc, but some require like one block from each validator during the epoch, including enough slack so spam cannot censor them.

This assumes the critical system paracahins cannot be tricked into wasting their own blocks, which likely becomes a risk only if the d-day governance supports smart contracts.

AndreiEres commented on polkadot/node/core/pvf/src/execute/queue.rs:
```diff
@@ -638,3 +653,299 @@ pub fn start(
+	///   approvals could not exceed 24%, even if there are no disputes.
+	/// - We cannot fully prioritize backing system parachains over backing other parachains based
+	///   on the distribution of the original 100%.
+	const PRIORITY_ALLOCATION_THRESHOLDS: &'static [(PvfExecPriority, usize)] = &[
+		(PvfExecPriority::Dispute, 70),
+		(PvfExecPriority::Approval, 80),
+		(PvfExecPriority::BackingSystemParas, 100),
+		(PvfExecPriority::Backing, 100),
+	];
```
That's a good question. I believe we'll have more information to make a decision after using it in production for a while.

eskimor commented on polkadot/node/subsystem-types/src/messages.rs:
```diff
@@ -204,6 +205,45 @@ pub enum CandidateValidationMessage {
+/// to separate and prioritize execution jobs by request type.
+/// The order is important, because we iterate through the values and assume it is going from higher
+/// to lowest priority.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, EnumIter)]
```
Given that we are already leaking to validation what we want to have validated here, we should likely keep the old name `PvfExecKind` and infer "priority" from this, together with other information. E.g. whether a validation is still worthwhile (ttl PR based on this one).
